File: 1034-subarrays-with-k-different-integers/subarrays-with-k-different-integers.py
class Solution:
    def solve(self,k:int,nums:List[int]):
        n=len(nums)
        l,r=0,0
        mp={}
        ans=0
        while r < n:
            mp[nums[r]]=mp.get(nums[r],0)+1
            while len(mp) > k:
                mp[nums[l]]-=1
                if mp[nums[l]]==0:
                    del mp[(nums[l])]
                l+=1
            ans+=(r-l+1)
            r+=1
        return ans
                


    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        n=len(nums)
        mp={}
        l,r=0,0
        ans=0
        # seen=set()
        # while r < n:
        #     mp[nums[r]]=mp.get(nums[r],0)+1
        #     seen.insert(nums[r])
        #     if len(seen) > k:
        #         mp[nums[l]]-=1
        #         if mp[nums[l]]==0:
        #             mp.delete(nums[l])
        #             seen.remove(nums[l]);
        #         l+=1
        #     if len(seen)==k:
        #         ans+=1
        #     r+=1
        kminus=self.solve(k-1,nums)
        kth=self.solve(k,nums)
        ans=kth-kminus
        return ans
                

File: 1034-subarrays-with-k-different-integers/subarrays-with-k-different-integers.cpp
class Solution {
public:
    int subarraysWithKDistinct(vector<int>& nums, int k) {
        int n=nums.size();
        unordered_map<int,int> mp1,mp2;
        int l=0,r=0;
        int ans_k=0;
        while(  r < n){
            mp1[nums[r]]+=1;
            while(mp1.size() > k){
                mp1[nums[l]]--;
                if(mp1[nums[l]] ==0) mp1.erase(nums[l]);
                l++;
            }
            ans_k+=(r-l+1);
            r++;
        }
        l=0,r=0;
        int ans_l=0;
        while(  r < n){
            mp2[nums[r]]+=1;
            while(mp2.size() > (k-1) ){
                mp2[nums[l]]--;
                if(mp2[nums[l]] ==0) mp2.erase(nums[l]);
                l++;
            }
            ans_l+=(r-l+1);
            r++;
        }
        return (ans_k - ans_l);
    }
};